File: leetcode/Easy/121. Best Time to Buy and Sell Stock/solution.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/submissions/1531156549
class Solution {
    public int maxProfit(int[] prices) {

        if (prices.length < 2) { return 0; }

        // sliding window for greedy solution
        int max = -99999999;
        int l, r;
        l = 0;
        r = 1;
        do {
            max = Math.max(max, Math.max(0, prices[r] - prices[l]));
            if (prices[r] < prices[l]) {
                l = r;
            }
            r++;
        } while (r <= prices.length - 1);
        return max;
    }
}

File: leetcode/Easy/121. Best Time to Buy and Sell Stock/solution.py
"""
--- Description ---

You are given an array prices where prices[i] is the price of a given stock on the ith day.

You want to maximize your profit by choosing a single day to buy one stock and choosing a different 
day in the future to sell that stock.

Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
"""

from math import inf


class Solution:
    def maxProfit(self, prices: list[int]) -> int:

        profit = 0
        low = inf

        for n in prices:
            if n < low:
                low = n
            if n - low > profit:
                profit = n - low
        
        return profit

"""
--- Submission ---

Runtime: 954 ms, faster than 99.36% of Python3 online submissions for Best Time to Buy and Sell Stock.
Memory Usage: 25.2 MB, less than 6.82% of Python3 online submissions for Best Time to Buy and Sell Stock.
"""